File: Civitas ruby/lib/mazo_sorpresas.rb
module Civitas
class MazoSorpresas
  
  def initialize(d=false)
    @barajada = false
    @usadas = 0
    @debug = d
    @sorpresas = []
    @cartasEspeciales = []
    @ultimaSorpresa = nil
    
    if(@debug)
      Diario.instance.object_id.ocurre_evento("El modo debug de mazo sorpresas se pone a 1")
    end
  end
  
  
  
  public
  
  def alMazo(s)
    if(!@barajada)
      @sorpresas.push(s)
    end
  end
  
  def siguiente()
    if(!@barajado || @sorpresas.length == @usadas)
      if(!@debug)
        @sorpresas = @sorpresas.shuffle
        @usadas = 0
        @barajada = true
      end
    end
    
    @usadas = @usadas + 1
    temporal = @sorpresas[0]
    @sorpresas.delete_at(0)
    @sorpresas.push(temporal)
    @ultimaSorpresa = temporal
        
    return temporal
  end
  
  def inhabilitarCartaEspecial(sorpresa)
    encontrado = @sorpresas.find_index(sorpresa)
    
    if(encontrado)
      carta = @sorpresas[encontrado]
      @sorpresas.delete_at(encontrado)
      @cartasEspeciales.push(carta)
      
      Diario.instance.object_id.ocurre_evento("Se inhabilita una carta especial")
    end
  end
  
  def habilitarCartaEspecial(sorpresa)
    encontrado = @cartasEspeciales.find_index(sorpresa)
    
    if(encontrado)
      carta = @cartasEspeciales[encontrado]
      @cartasEspeciales.delete_at[encontrado]
      @sorpresas.push(carta)
      
      Diario.instance.object_id.ocurre_evento("Se habilita una carta especial")
    end
  end  
end
end

File: Civitas java/src/civitas/MazoSorpresas.java
package civitas;
import java.util.ArrayList;
import java.util.Collections;

public class MazoSorpresas{
    ArrayList<Sorpresa> sorpresas;
    boolean barajada;
    int usadas;
    boolean debug;
    ArrayList<Sorpresa> cartasEspeciales;
    Sorpresa ultimaSorpresa;

    private void init(){
        barajada = false;
        usadas = 0;
        sorpresas = new ArrayList();
        cartasEspeciales = new ArrayList();
    }

    MazoSorpresas(boolean valor){
        debug = valor;
        init();
        if(debug)
            (Diario.getInstance()).ocurreEvento("Se modifica el estado debug de MazoSorpresas a" + debug);
    }

    MazoSorpresas(){
        debug = false;
        init();
    }

    void alMazo(Sorpresa s){
        if(!barajada)
            sorpresas.add(s);
    }

    Sorpresa siguiente(){
        if(!barajada || usadas > sorpresas.size()){
            if(!debug){
                Collections.shuffle(sorpresas);
                usadas = 0;
                barajada = true;
                usadas++;
            }
        }
        
        Sorpresa temporal = sorpresas.get(0);
        ultimaSorpresa = temporal;
        sorpresas.remove(0);
        sorpresas.add(temporal);

        return temporal;
    }

    void inhabilitarCartaEspecial(Sorpresa sorpresa){
        boolean contiene;
        contiene = sorpresas.contains(sorpresa);

        if(contiene){
            int indice;
            Sorpresa carta;
            indice = sorpresas.indexOf(sorpresa);
            carta = sorpresas.get(indice);
            sorpresas.remove(sorpresa);
            cartasEspeciales.add(carta);

            (Diario.getInstance()).ocurreEvento("Se inhabilita una carta especial");
        }
    }

    void habilitarCartaEspecial(Sorpresa sorpresa){
        boolean contiene;
        contiene = cartasEspeciales.contains(sorpresa);

        if(contiene){
            int indice;
            Sorpresa carta;

            indice = cartasEspeciales.indexOf(sorpresa);
            carta = cartasEspeciales.get(indice);
            sorpresas.add(carta);

            (Diario.getInstance()).ocurreEvento("Se habilita una carta especial");
        }
    }
}
